packages/optimizer/lib/transformers/AutoExtensionImporter.js

```diff
@@ -0,0 +1,288 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+
+const {findMetaViewport} = require('../HtmlDomHelper');
+const {calculateHost} = require('../RuntimeHostHelper');
+
+const BIND_SHORT_FORM_PREFIX = 'bind';
+const AMP_BIND_DATA_ATTRIBUTE_PREFIX = 'data-amp-bind-';
+
+// Some AMP component don't bring their own tag, but enable new attributes on other
+// elements. Most are included in the AMP validation rules, but some are not. These
+// need to be defined manually here.
+const manualAttributeToExtensionMapping = new Map([
+  ['mask', 'amp-inputmask'],
+  ['lightbox', 'amp-lightbox-gallery'],
+]);
+
+/**
+ * Extension Auto Importer - this transformer auto imports all missing AMP extensions.
+ *
+ * The importer analyzes the HTML source code and identifies missing AMP extension imports
+ * using multiple strategies:
+ *
+ * - use validation rules to map used AMP tags to required AMP extensions.
+ * - use validation rules to map used AMP attributes to required AMP extensions.
+ * - manually specifiy attribute to extension mappings if this information is not available in the
+ *   validation rules.
+ * - mnullay implement AMP extension detection for a few corner cases.
+ *
+ * This importer also enables a shortcode `bindtext` instead of `data-amp-bind-text` for specifying
+ * AMP bindings when the square bracket notation (`[text]`) is not available. To avoid accidently
+ * rewriting non-AMP attributes, the transformer uses the AMP validation rules to only rename bindable
+ * attributes as specified in the validation rules.
+ *
+ * You can disable the auto extension import by passing `{ autoExtensionImport: false }` via the config.
+ */
+class AutoExtensionImporter {
+  constructor(config) {
+    this.enabled = config.autoExtensionImport !== false;
+    this.log_ = config.log.tag('AutoExtensionImporter');
+
+    // We use the validation rules to infer extension imports. The rules are downloaded once and for
+    // efficency, we initially extract all needed rules
+    this.initExtensionSpec_(config.validatorRules);
+  }
+
+  /**
+   * @private
+   */
+  async initExtensionSpec_(validatorRules) {
+    this.extensionSpec = validatorRules.fetch().then((rules) => {
+      // Map extension names to more info
+      const extensionsMap = new Map(rules.extensions.map((ext) => [ext.name, {
```

Comment:
Thinking about this... you may want to take `htmlFormat` into account. For example, AMP for Email doesn't support `amp-carousel` 0.2.

More importantly, this assumes there's exactly one extension for a given `ext.name` which is incorrect because of the above. I'm not sure what the current behavior is, but it will give you an arbitrary `amp-carousel` that may be the email one (and thus 0.1) or the regular one (and thus 0.1 and 0.2).

Reply:
Yup. That's already on my todo list.